functions/src/userManagement.ts

```diff
@@ -0,0 +1,25 @@
+import * as functions from 'firebase-functions'
+import * as admin from 'firebase-admin'
+import { storageBucketMock, firestoreMock } from './test/mocks'
+import Refs from './firestoreRefs'
+import * as _ from 'lodash'
+
+try { admin.initializeApp() } catch (e) {}
+
+const isTestMode = process.env.NODE_ENV === 'test'
+const database = isTestMode ? firestoreMock : admin.firestore()
+const storageBucket = isTestMode ? storageBucketMock : admin.storage().bucket()
```

Comment:
I still think that I would prefer we wrap the whole thing in a function that takes these dependencies as an argument as I have the following issues with this:
1. It feels like the Volkswagen trick of looking one way during emissions testing and another normally...
2. It necessitates us deploying our test code to firebase - you're not really supposed to deploy test code.
3. It makes it so there is no way for us to perform integration testing - I know we're not currently, but to remove the option to use a real database seems like an oversite.



I'm not saying it NEEDS to be fixed, but this is generally considered bad practice.

Reply:
Just to get the conversation on github:
1. I don't quite track with this analogy - the car itself remains the same each time and doesn't itself make any changes just to pass the tests. Practically (if not technically) speaking, this is exactly what we accomplish with dependency injection, just at the root of the file instead of specifying it in every function.
2. 100% onboard with this, and I'd love to avoid it
3. This is fair. I think a big concern I have is DX - I'm not interested in specifying which database should be used for EVERY single test.

What would be a practical solution? Do you think going through something like [proxyquire](https://github.com/thlorenz/proxyquire) would solve this better? One extremely nice thing about the current setup is that it's insanely easy to add new stubs and to spy on anything about the mocked API. I'm not sure if it's worth it to sacrifice that if it means a _ton_ more setup and configuration for our tests.

In this particular case, I was copy/pasting the configuration from previous tests, so this is nothing new. ;)

I think that a re-architecture of this dependency injection would definitely be helpful, but I think that we should dedicate a PR to that specifically since it touches on more than just these files. I think we could also put effort in to not use `// @ts-ignore` and instead make the mocks more typescript-y. As of now, they're way more javascript-y and take advantage of the fact that everything in javascript is just words  and that it doesn't yell at you for not exactly matching the type. Obviously a mock will never exactly match the functionality or type of the actual API :P